### PR TITLE
Feature/model reader writer

### PR DESCRIPTION
**Pull request by m-nash**
Adds public APIs to be able to read and write models in multiple formats.

Replacement for https://github.com/Azure/azure-sdk-for-net/pull/38506

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/core/Azure.Core/src/RequestContent.cs
Comment (m-nash) [inline sdk/core/Azure.Core/src/RequestContent.cs:132]: @annelo-msft I don't think this is how we want to do this longer term but it works for an initial draft to get the alpha out.

File: sdk/core/System.Net.ClientModel/src/Experimental/ModelSerializerHelper.cs
Comment (annelo-msft) [inline sdk/core/System.Net.ClientModel/src/Experimental/ModelSerializerHelper.cs:7]: +1 to putting this in Internal!

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/IModel.cs
Comment (annelo-msft) [inline sdk/core/System.Net.ClientModel/src/ModelReaderWriter/IModel.cs:11]: I defer to @KrzysztofCwalina on where we want to land with names prior to the review with the BCL team and feedback we might get with them, but I wanted to document some observations I have (non-blocking for this PR):
- Could the `IModel` name be confused with the package name `ClientModel`?  
  - Questions on this might include: Why are we modifying the name `Model` in the package name but not in this interface name?
- What does the word "model" mean in BCL APIs like `ClientModel` and `ServiceModel` and elsewhere, and is this use of the word model consistent with that vocabulary and meaning?
- Where we also have IJsonModel - is the model itself about Json?  Why is `Json` the best modifier for `Model`?

One alternative proposal I might toss out there as a starting point for conversation -- and motivated by the observation that .NET interfaces are often used to describe the behaviors enable (per .NET not supporting multiple inheritance), i.e. IEnumerable and IDisposable vs. the "is-a" style interfaces like IList and ICollection -- would be what about something like `IContentReadWriteable` and `IJsonContentReadWriteable`?

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs
Comment (annelo-msft) [inline sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs:13]: Another non-PR-blocking question/comment, more as questions to prep for BCL review...  Is there a need for this type to do both reading and writing?  What is it?  I'm wondering, given the at-first-glance niceness of interface names like `IModelReadable` and `IModelWriteable` how disruptive/costly to the implementation it would be to split components for reading and writing to maybe simplify the APIs?  I feel like we often get developer and partner questions like "why didn't my model get a serialize method generated?"  For an output-only model, we don't tend to generate the serialize methods, so will coupling read and write mean that we are changing team precedent regarding that aspect of today's model design?  If we split them, output-only models could feasibly only implement the `IModelReadable` interface, and it might be more clear what is happening there and why...

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs
Comment (annelo-msft) [inline sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs:14]: Do we know if there is a word that maps better to the BCL lexicon than `Wire`?

There is a nice article that maps out BCL terminology in this space that I spent some time with recently here: https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient  It was very helpful in getting my head around how BCL things about HTTP network operations and what words they use to describe the various aspects of that.

File: sdk/core/System.Net.ClientModel/src/Properties/AssemblyInfo.cs
Comment (annelo-msft) [inline sdk/core/System.Net.ClientModel/src/Properties/AssemblyInfo.cs:7]: Thanks for fixing this!

File: sdk/core/System.Net.ClientModel/tests/internal/ModelReaderWriter/ModelJsonConverterTests.cs
Comment (annelo-msft) [inline sdk/core/System.Net.ClientModel/tests/internal/ModelReaderWriter/ModelJsonConverterTests.cs:18]: Also non-blocking given timelines, but curious - if ModelJsonConverter is in the public API, would it be possible to move it into the System.Net.ClientModel.Tests project instead of internal?  Before we merge this PR to main, I would like to be a little more rigorous about only having tests in System.Net.ClientModel.Tests.Internal that test internal APIs and have tests for public APIs in the main project.  

If this is properly a test of an internal API, feel free to let me know that that is the case and that I'm just not seeing it yet.

File: sdk/core/Azure.Core/api/Azure.Core.net461.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/api/Azure.Core.net461.cs:545]: Per offline discussion, I would like to remove this API prior to review with the BCL team and am looking to see if I can spot an easy way to do it now, and will report back.

File: sdk/core/Azure.Core/tests/ModelReaderWriter/ModelReaderWriterRequestContentTests.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/ModelReaderWriter/ModelReaderWriterRequestContentTests.cs:28]: What is the reason that we need a RequestContent here instead of a PipelineContent?

File: sdk/core/Azure.Core/tests/ModelReaderWriter/ModelReaderWriterRequestContentTests.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/ModelReaderWriter/ModelReaderWriterRequestContentTests.cs:63]: Same question here regarding why we need to create an instance of PipelineContent to pass to RequetsContent.Create instead of passing the model directly?

File: sdk/core/Azure.Core/tests/common/ModelReaderWriter/Models/DiscriminatorSet/BaseModel.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/common/ModelReaderWriter/Models/DiscriminatorSet/BaseModel.cs:25]: I think this should be able to call a RequestContent.Create(IModel) method directly as well rather than creating a PipelineContent first.  This will simplify the pattern in the generated code.

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs
Comment (m-nash) [reply sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs:13]: We will always create both the Write method and Read method for all models going forward regardless of usage.  The reason for this is the use case for a customer is to be able to store the full state of the model.  The fact that the model might be output only or input only doesn't matter they will need both Read and Write.

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs
Comment (m-nash) [reply sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs:14]: Not that I know of do you have a suggestion?

Comment (annelo-msft) [reply sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs:14]: Not offhand, but I will think about it.  Mostly just flagging that this could come up in the BCL review.

**Review by annelo-msft — APPROVED** [review]
Approving with the assumption that:
- https://github.com/Azure/azure-sdk-for-net/pull/39421/files#r1367489853 will be addressed before merging.
- If time allows, it would be great to address https://github.com/Azure/azure-sdk-for-net/pull/39421#discussion_r1367472600 as well.  For this one, I'm interested in understanding how disruptive this request is, and a bit concerned that either we'll forget to come back to it, or open the feature branch up to a larger set of contributors who'll copy the pattern of putting all tests in the internal test project and lose the benefits of making the split, and/or give ourselves a lot of work educating and chasing to get the outcome we want by setting the precedent early that the internal test project is only for tests of internal APIs.

I'm fine to come back to the naming comments in a later PR.

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/IModel.cs
Comment (annelo-msft) [reply sdk/core/System.Net.ClientModel/src/ModelReaderWriter/IModel.cs:11]: One additional comment here is that the "is-a" interfaces like IList and ICollection have a lot of APIs that are required for those concrete types.  Adding a method to an interface after it GAs is a breaking change so by calling an interface with only `Read` and `Write` operations `IModel` sort of means "cloud client models only read and write their content".  Is there a possible future world where a cloud client model would have other behaviors we wanted to enforce via interfaces?

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs
Comment (annelo-msft) [reply sdk/core/System.Net.ClientModel/src/ModelReaderWriter/ModelReaderWriterFormat.cs:14]: Ok, I have an idea.

This is the one that is "whatever the service expects", is that correct?

If so, what about `ServiceContract`?

For me, "wire" communicates the transmission mechanism, e.g. "I send my message over the wire," and less about the piece that receives it on the other side of the wire.  It feels a bit like a throwback to land-line telephone communications.

`ServiceContract` feels like it communicates better the intent of "what the service expects."  It is the contract with the service.  I believe BCL uses the word service in the same way, so this is at least consistent there, and if they have a better word for "contract," this name implies enough that they could proactively tell us that in a review, without having to ask us to explain what "wire" means first.

I also think it goes better alongside "Json" in the enum values, because a customer looking at those side-by-side might wonder "hey, the service contract using JSON, so this other Json must mean something else," and might be able to logic it out themselves, and understand it completely with a boost from good refdoc comments.  Otherwise, it feels to me that people could get confused between whether to use "Wire" or "Json" when the service contract calls for JSON -- I know I was confused about that at first.

What do you think?

File: sdk/core/Azure.Core/api/Azure.Core.net461.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/api/Azure.Core.net461.cs:545]: 👍

File: sdk/core/System.Net.ClientModel/src/ModelReaderWriter/IModel.cs
Comment (annelo-msft) [reply sdk/core/System.Net.ClientModel/src/ModelReaderWriter/IModel.cs:11]: One other angle on this as a BCL type might be ... would any other non-model component ever want to use these behaviors as specified in the interface, and so we would want to name the interface to describe the behavior so that that would make sense?

**Pull request merged**
